### PR TITLE
Windows Port from Microsoft

### PR DESCRIPTION
**Pull request by yuslepukhin**
Summary: Make RocksDb build and run on Windows to be functionally
 complete and performant. All existing test cases run with no
 regressions. Performance numbers are in the pull-request.

 Test plan: make all of the existing unit tests pass, obtain perf numbers.

 Co-authored-by: Praveen Rao praveensinghrao@outlook.com
 Co-authored-by: Sherlock Huang baihan.huang@gmail.com
 Co-authored-by: Alex Zinoviev alexander.zinoviev@me.com
 Co-authored-by: Dmitri Smirnov dmitrism@microsoft.com

### CODE REVIEW

**Comment by igorcanadi** [comment]
Wow, this is amazing! Thanks a lot for doing this!

Would it be possible to split this pull request into a series of a lot of smaller PRs? For example, one PR could be adding ROCKSDB_LIBRARY_API into C API, another could be adding ROCKSDB_PRIszt instead of %zu, etc. That will make the review process much easier.

Also, once we get RocksDB working on Windows, would it be possible to setup some continuous builds on your side? We want to make sure we keep it working :)

File: db/c.cc [hunk 1]
```diff
@@ -12,7 +12,7 @@
 #include "rocksdb/c.h"
```
Comment (yuslepukhin) [inline db/c.cc]: Lost some newer functions within the PR. Revisit.

**Comment by yuslepukhin** [comment]
Igor, we are trying to evaluate the effort involved into different parts of your request.
This morning this PR looks a lot more deficient than yesterday.
Continuous integration is a desirable feature.
For the benefit of interested folks, this port is functional and is targeted for production use within MS.

File: db/db_bench.cc [hunk 2]
```diff
@@ -926,23 +900,9 @@ struct DBWithColumnFamilies {
...
-    cfh.clear();
-    if (txn_db) {
-      delete txn_db;
-      txn_db = nullptr;
-    } else {
-      delete db;
-    }
-    db = nullptr;
-  }
-
```
Comment (siying) [inline db/db_bench.cc]: I noticed quite some functions were deleted. Are they repository sync and merge issue?

Comment (yuslepukhin) [inline db/db_bench.cc]: Yes, this diff is deficient. There are other omissions that I noticed.We have a new one coming.

File: build_tools/runall.bat [hunk 3]
```diff
@@ -0,0 +1,99 @@
+@echo off
```
Comment (yuslepukhin) [inline build_tools/runall.bat]: This will will go away

File: port/win/env_win.cc [hunk 4]
```diff
@@ -0,0 +1,2604 @@
...
+// Roundup x to a multiple of y
+inline
+size_t Roundup(size_t x, size_t y) {
+  return ((x + y - 1) / y) * y;
+}
+
+
+// Can only truncate or reserve to a sector size aligned if
+// used on files that are opened with Unbuffered I/O
+// Normally it does not present a problem since in memory mapped files
```
Comment (yuslepukhin) [inline port/win/env_win.cc]: Comment is no longer true.

**Comment by siying** [comment]
Need to update INSTALL.md to introduce how to build it in Windows?

**Comment by yuslepukhin** [comment]
Sying, current instructions are inside the CMakeLists.txt which is the CMake file. We will add those to the INSTALL.md per convention.

**Comment by siying** [comment]
@yuslepukhin sure. Just mention in the INSTALL.md, if to build in Windows, follow instruction in CMakeLists.txt. It will be great.

File: db/file_indexer.h [hunk 5]
```diff
@@ -58,7 +58,8 @@ class FileIndexer {
                    std::vector<FileMetaData*>* const files);
 
   enum {
-    kLevelMaxIndex = std::numeric_limits<int32_t>::max()
+    // MSVC version 1800 still does not have constexpr for ::max()
+    kLevelMaxIndex = INT32_MAX
```
Comment (siying) [inline db/file_indexer.h]: Can you create a const in port.h and still use std::numeric_limits<int32_t>::max() for other platforms? I'm concerned some platform might not have INT32_MAX.

File: db/managed_iterator.cc [hunk 6]
```diff
@@ -93,6 +93,7 @@ ManagedIterator::~ManagedIterator() {
     snapshot_created_ = false;
     read_options_.snapshot = nullptr;
   }
+  UnLock();
```
Comment (siying) [inline db/managed_iterator.cc]: Not getting this change.

File: include/rocksdb/slice.h [hunk 7]
```diff
@@ -24,6 +24,12 @@
 #include <stddef.h>
 #include <string.h>
 #include <string>
+#include <stdio.h>
+
+// Do not want to include the whole /port/port.h here for one define
+#ifdef OS_WIN
+#  define snprintf _snprintf
+#endif
```
Comment (siying) [inline include/rocksdb/slice.h]: Instead, can we move implementation of ToString() to util/slice.cc?

File: db/managed_iterator.cc (hunk 6 above)
Comment (yuslepukhin) [inline db/managed_iterator.cc]: Debug diagnostics was (correctly) complaining that the mutex is being destroyed while owned because it is acquired at the start of the __dtor. The proper fix would be to remove the Lock() line since the object must not be ordinarily a subject of MT access at the time of destruction but we apparently decided to play along in this case and simply unlocked it.

File: include/rocksdb/thread_status.h [hunk 8]
```diff
@@ -31,7 +32,15 @@ namespace rocksdb {
...
 // TODO(yhchiang): remove this function once c++14 is available
 //                 as std::max will be able to cover this.
+#ifndef OS_WIN
 constexpr int constexpr_max(int a, int b) { return a > b ? a : b; }
+#else
+// Current MS compiler does not support constexpr
+template<int A, int B>
+struct constexpr_max {
+  static const int result = (A > B) ? A : B;
+};
```
Comment (siying) [inline include/rocksdb/thread_status.h]: Maybe replace existing codes this template based one.

**Comment by siying** [comment]
I'm not an expert of make systems, but is there a way for CMakeLists.txt to use src.mk, instead of duplicating the list?

**Comment by maykov** [comment]
In MyRocks, we get a list of files from src.mk in a somewhat hacky way. See https://github.com/MySQLOnRocksDB/mysql-5.6/blob/webscalesql-5.6.23.71/storage/rocksdb/get_rocksdb_files.sh

File: port/win/env_win.cc [hunk 9]
```diff
@@ -0,0 +1,2604 @@
...
+          s = IOError("Failed to fdopen NewSequentialFile: " + fname, code);
+        } else {
+          result->reset(new WinSequentialFile(fname, file, options));
+        }
+      }
+    }
+    return s;
+  }
+
+  virtual Status NewRandomAccessFile(const std::string& fname, std::unique_ptr<RandomAccessFile>* result, const EnvOptions& options) override {
```
Comment (siying) [inline port/win/env_win.cc]: We follow Google C++ style: https://google-styleguide.googlecode.com/svn/trunk/cppguide.html and we use 80-char wrapping.
We have a tool build_tools/format-diff.sh to auto reformatting. Not sure whether you can run it on Windows. Here are the auto-formatting patch: https://reviews.facebook.net/F515421

**Comment by siying** [comment]
@yuslepukhin  any idea this message means?

CMake Error at CMakeLists.txt:33 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.

File: db/file_indexer.h (hunk 5 above)
Comment (yuslepukhin) [inline db/file_indexer.h]: @siying Let's chat quickly after you done with the review. I sent you personal email.

**Comment by siying** [comment]
@yuslepukhin I think I figured out the problem of the failure. Please mention in CMakeLists.txt that command line tool "git" needs to be available to run (in PATH) while running cmake.

**Comment by zinoale** [comment]
I think the best way would be to merge both build systems (make/Linux and cmake/Windows) into a single one and thus have no duplication for the list.

On Jul 06, 2015, at 12:11 PM, Siying Dong <notifications@github.com> wrote:

I'm not an expert of make systems, but is there a way for CMakeLists.txt to use src.mk, instead of duplicating the list?

File: db/file_indexer.h (hunk 5 above)
Comment (yuslepukhin) [inline db/file_indexer.h]: @siying INT32_MAX is not in limits.h and may not be in stdint.h so it is fare. Do you want a similar approach for other constants or just this one?

Comment (siying) [inline db/file_indexer.h]: @yuslepukhin It's better to be consistent but I didn't yet see another one yet.

File: db/managed_iterator.cc (hunk 6 above)
Comment (siying) [inline db/managed_iterator.cc]: Oh it is a bug then. Great.

File: table/format.h [hunk 10]
```diff
@@ -191,6 +191,18 @@ struct BlockContents {
...
+  BlockContents(BlockContents&& other) {
+      *this = std::move(other);
+  }
+
+  BlockContents& operator=(BlockContents&& other) {
+      data = std::move(other.data);
+      cachable = other.cachable;
+      compression_type = other.compression_type;
+      allocation = std::move(other.allocation);
+      return *this;
```
Comment (siying) [inline table/format.h]: We use 2-char space indenting.

File: table/plain_table_reader.cc [hunk 11]
```diff
@@ -114,13 +114,11 @@ PlainTableReader::~PlainTableReader() {
...
                               const EnvOptions& env_options,
+                              const PlainTableOptions& table_options,
                               const InternalKeyComparator& internal_comparator,
-                              unique_ptr<RandomAccessFile>&& file,
+                              std::unique_ptr<RandomAccessFile>&& file,
                               uint64_t file_size,
-                              unique_ptr<TableReader>* table_reader,
-                              const int bloom_bits_per_key,
-                              double hash_table_ratio, size_t index_sparseness,
-                              size_t huge_page_tlb_size, bool full_scan_mode) {
```
Comment (siying) [inline table/plain_table_reader.cc]: I like the change of passing through PlainTableOptions all the way down. But is it related to the port?

File: util/arena.cc [hunk 12]
```diff
@@ -52,12 +54,15 @@ Arena::~Arena() {
   for (const auto& block : blocks_) {
     delete[] block;
   }
+// yuslepukhin: this needs to be addressed as it previously was under #ifdef
+#ifndef OS_WIN
```
Comment (siying) [inline util/arena.cc]: #ifdef MAP_HUGETLB
should be enough.

File: table/plain_table_reader.cc (hunk 11 above)
Comment (yuslepukhin) [inline table/plain_table_reader.cc]: This is a new feature that we needed for integrating this into our database.  Allows to pass a config string and get options instance back. I think this is the only one here.

File: util/dynamic_bloom.h [hunk 13]
```diff
@@ -9,7 +9,8 @@
 
 #include "rocksdb/slice.h"
 
-#include "port/port_posix.h"
+
+#include <port/port.h>
```
Comment (siying) [inline util/dynamic_bloom.h]: Is there a reason changing from ".." to <..>?

File: util/options_helper.cc [hunk 14]
```diff
@@ -389,10 +394,29 @@ bool ParseColumnFamilyOption(const std::string& name, const std::string& value,
...
+      Status mem_factory_s = 
+          GetMemTableRepFactoryFromString(value, &new_mem_factory);
+      if (!mem_factory_s.ok()) {
+        return false;
+      }
+      new_options->memtable_factory.reset(new_mem_factory);
     } else if (name == "min_write_buffer_number_to_merge") {
       new_options->min_write_buffer_number_to_merge = ParseInt(value);
-    } else if (name == "max_write_buffer_number_to_maintain") {
-      new_options->max_write_buffer_number_to_maintain = ParseInt(value);
```
Comment (siying) [inline util/options_helper.cc]: Is it a merge error?

File: util/options_helper.cc [hunk 15]
```diff
@@ -648,6 +672,117 @@ Status GetBlockBasedTableOptionsFromString(
...
+    return Status::InvalidArgument("cuckoo is not supported for now");
+    // TODO(bahuang): cuckoo is not supported for now
+    // *new_mem_factory = NewHashCuckooRepFactory(
+    //    options.write_buffer_size, FLAGS_key_size + FLAGS_value_size));    
+  } else {
+    return Status::InvalidArgument("Can't parse option " + opts_str);
+  }
+  return Status::OK();
+}
+
```
Comment (siying) [inline util/options_helper.cc]: Are these changes related? Can we separate them if they are not related.

File: util/arena.cc [hunk 16]
```diff
@@ -8,13 +8,15 @@
...
 #include <sys/mman.h>
+#endif
+#include "port/port.h"
 #include <algorithm>
 #include "rocksdb/env.h"
 
 namespace rocksdb {
 
-const size_t Arena::kInlineSize;
```
Comment (siying) [inline util/arena.cc]: This is used by ./util/arena_test.cc and test breaks for it.

Comment (siying) [inline util/arena.cc]: After removing it, it does not build under gcc anymore. Can you add it back if it is not Windows?

File: util/thread_local.cc [hunk 17]
```diff
@@ -15,10 +15,89 @@
...
+void NTAPI WinOnThreadExit(PVOID module, DWORD reason, PVOID reserved) {
+    // We decided to punt on PROCESS_EXIT
+    if (DLL_THREAD_DETACH == reason) {
+        if (thread_local_key != -1 && thread_local_inclass_routine != nullptr) {
+            void* tls = pthread_getspecific(thread_local_key);
+            if(tls != nullptr) {
+                thread_local_inclass_routine(tls);
+            }
+        }
+     }
```
Comment (siying) [inline util/thread_local.cc]: 2-char indenting.

File: port/win/env_win.cc (hunk 9 above)
Comment (yuslepukhin) [inline port/win/env_win.cc]: @siying Cygwin bash refuses to run format-diff for some bizarre reason. 

$ build_tools/format-diff.sh
build_tools/format-diff.sh: line 108: syntax error: unexpected end of file

Would you be fine if I applied this patch on top on my most recent internal code review changes? I am asking because the patch includes some posix only files and others we did not change and they are not related to this port.

Comment (siying) [inline port/win/env_win.cc]: It's fine. Or you can first check it in and I fix it in master quickly in another patch.

**Comment by siying** [comment]
Thank you for the great patch! It mostly looks good to me. Need to improve documents about build instruction, especially how to install/disable the dependencies. Need to fix kInlineSize, which is breaking the build using gcc. Some merge error needs to be fixed.

I'm not able to review codes for OS_WIN. I will leave to the community to comment on it.

**Comment by siying** [comment]
env_win.cc, line 972 "assert(!options.use_mmap_reads);" fails on my 32-bit laptop, because mmap is chosen to be not executed:
if (options.use_mmap_reads && sizeof(void*) >= 8) {
}
Need to fix this assert.

**Comment by yuslepukhin** [comment]
Will in the documentation that this is a 64-bit only port.

**Comment by siying** [comment]
@yuslepukhin that is fine.

**Comment by yuslepukhin** [comment]
@igorcanadi We are actively exploring ways to setup a CI.

**Comment by yuslepukhin** [comment]
@siying I think this represents the latest we would like to check in unless others have more comments. We have addressed all but one. We still do not have a single list of source files.

**Comment by siying** [comment]
@yuslepukhin  that is fine. We can address it later. I'll take a look and merge it if it is generally OK. I don't want to hold big patch for too long.

File: db/db_impl.cc [hunk 18]
```diff
@@ -4151,7 +4151,7 @@ Status DBImpl::GetDbIdentity(std::string& identity) const {
   if (!s.ok()) {
     return s;
   }
-  char buffer[file_size];
+  char* buffer = reinterpret_cast<char*>(alloca(file_size));
```
Comment (igorcanadi) [inline db/db_impl.cc]: I think this will fail valgrind. Can you please use unique_ptr for this?

File: db/db_test.cc [hunk 19]
```diff
@@ -7,10 +7,16 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file. See the AUTHORS file for names of contributors.
 
+// Introduction of SyncPoint effectively disabled building and running this test in Release build.
+// which is a pity, it is a good test
+#if !(defined NDEBUG) || !defined (OS_WIN)
```
Comment (igorcanadi) [inline db/db_test.cc]: We don't usually run our tests in release build.

File: db/db_impl.cc (hunk 18 above)
Comment (yuslepukhin) [inline db/db_impl.cc]: @igorcanadi sure. Pls, let me know when you are done with the review.

File: db/db_test.cc (hunk 19 above)
Comment (yuslepukhin) [inline db/db_test.cc]: @igorcanadi Understood. However, just FYI, on Windows Release build perf diff is about 10x with Debug and that affects perf related tests. Also on one occasion we had test pass in Debug but not in Release.

**Comment by yuslepukhin** [comment]
@igorcanadi We are thinking along these lines. https://msdn.microsoft.com/en-us/Library/vs/alm/Build/github/index

File: db/db_impl.cc (hunk 18 above)
Comment (siying) [inline db/db_impl.cc]: @igorcanadi I did ran valgrind for their previous version and didn't see any issue. I think this is not new codes.

File: db/db_test.cc [hunk 20]
```diff
@@ -8810,7 +8825,7 @@ class RecoveryTestHelper {
     ASSERT_GT(fd, 0);
     ASSERT_EQ(offset, lseek(fd, offset, SEEK_SET));
 
-    char buf[len];
+    void* buf = alloca(len);
```
Comment (igorcanadi) [inline db/db_test.cc]: same here, i think this leaks memory. not sure if valgrind works on windows, but it might be good to run it, if it does.

File: port/win/env_win.cc (hunk 9 above)
Comment (yuslepukhin) [inline port/win/env_win.cc]: @siying After fixing the line endings and installing python scripts and modules, this thing starts but python dumps core. I will rely on you to fix the formatting while I will see to produce a properly formatted code to begin with. GCC, however, runs fine is valuable.

File: include/rocksdb/transaction_log.h [hunk 21]
```diff
@@ -56,8 +56,25 @@ class LogFile {
...
-  std::unique_ptr<WriteBatch> writeBatchPtr;
+    SequenceNumber sequence = 0;
+    std::unique_ptr<WriteBatch> writeBatchPtr;
+
+    BatchResult() {
+    }
+
+    BatchResult(const BatchResult&) = delete;
+
+    BatchResult& operator=(const BatchResult&) = delete;
```
Comment (igorcanadi) [inline include/rocksdb/transaction_log.h]: Rule of five requires us to declare either all of {copy, copy-assign, move, move-assign, destructor}, or none of them. See here: http://en.cppreference.com/w/cpp/language/rule_of_three

File: port/win/port_win.h [hunk 22]
```diff
@@ -0,0 +1,576 @@
...
+    // it does NOT verify that mutex is held by a calling thread
+    void AssertHeld();
+
+    std::unique_lock<std::mutex>& getLock() {
+        return lock;
+    }
+
+private:
+    friend class CondVar;
+    std::mutex m_mutex;
```
Comment (igorcanadi) [inline port/win/port_win.h]: Feel free to replace all of our Mutex and CondVar classes with C++11 equivalent (even in posix). However, you don't need to do it as part of this patch. We can do it later.

File: db/db_test.cc (hunk 20 above)
Comment (yuslepukhin) [inline db/db_test.cc]: @igorcanadi. Yes, this was in previous diff. This will not leak as this is a rough equivalent of reserving a stack space in a dynamic manner. For this reason, if valgrind does not complain, I'd advocate leaving this as is because it would be close to the original code perf wise.

File: port/win/port_win.h [hunk 23]
```diff
@@ -0,0 +1,576 @@
...
+private:
+    std::condition_variable cv_;
+    Mutex * mu_;
+};
+
+typedef std::once_flag OnceType;
+#define LEVELDB_ONCE_INIT std::once_flag::once_flag();
+extern void InitOnce(OnceType* once, void (*initializer)());
+
+inline bool Snappy_Compress(const CompressionOptions& opts, const char* input,
```
Comment (igorcanadi) [inline port/win/port_win.h]: I'm curious why do you need a separate implementation of compression functions? I thought they're platform independent.

File: include/rocksdb/transaction_log.h (hunk 21 above)
Comment (yuslepukhin) [inline include/rocksdb/transaction_log.h]: @igorcanadi Will add a destructor

File: third-party/fbson/FbsonDocument.h [hunk 24]
```diff
@@ -125,9 +125,12 @@ class FbsonDocument {
     uint8_t ver_;
   } header_;
 
-  char payload_[0];
+  char payload_[1];
```
Comment (igorcanadi) [inline third-party/fbson/FbsonDocument.h]: At some point it would be good to contribute this back to MySQL project, since this file was copied from there. Please add a TODO into the COMMIT.md file in this directory.

File: port/win/port_win.h (hunk 23 above)
Comment (yuslepukhin) [inline port/win/port_win.h]: @igorcanadi This is an oversight, probably a copy paste from the original porting effort months ago.

File: port/win/port_win.h (hunk 22 above)
Comment (siying) [inline port/win/port_win.h]: I would hold for this for now. There is a possibility that Natham wants to implement our own mutex to improve write queue sychronization.

File: include/rocksdb/transaction_log.h (hunk 21 above)
Comment (igorcanadi) [inline include/rocksdb/transaction_log.h]: You also need to define copy constructor and copy assignment, right?

File: db/db_test.cc (hunk 20 above)
Comment (igorcanadi) [inline db/db_test.cc]: Ah, `alloca` gets bytes from stack, got it. This is fine then. db_test perf is not very critical :)

File: db/db_impl.cc (hunk 18 above)
Comment (igorcanadi) [inline db/db_impl.cc]: alloca takes from stack, not heap. so this is not an issue.

**Comment by igorcanadi** [comment]
I just took a quick look because I assume @siying already reviewed this in detail. LGTM.

File: include/rocksdb/transaction_log.h (hunk 21 above)
Comment (yuslepukhin) [inline include/rocksdb/transaction_log.h]: @igorcanadi Not really, the semantics of the object and how it was and being used does not suggest that we need two functional copies of it due to the fact that the std::unique_ptr also has copy and assignment deleted. So I chose to replicate those because it yields fast return from functions and container insertion automatically w/o std::move.

File: db/db_impl.cc (hunk 18 above)
Comment (siying) [inline db/db_impl.cc]: I reran valgrind for DBTest.IdentityAcrossRestarts and it passed.

File: port/win/port_win.h (hunk 23 above)
Comment (yuslepukhin) [inline port/win/port_win.h]: @igorcanadi I will have to address this separately.

**Comment by siying** [comment]
I'm going to merge it. Any time we see, we'll fix it as a follow-up. Next week is our TechDebt week. If we hold this patch to next week, there will be lots of rebasing/retesting works to do.

If people have any comments, feel free to continue commenting here. We'll actively fix them after the merge.

**Comment by siying** [comment]
As discussed offline. I'm running a "make format" against this patch and sending a code review for it: https://reviews.facebook.net/D41961

**Comment by yuslepukhin** [comment]
Current build is broken. I am moving port/port.h includes out of public headers. Perhaps, we could format after that?

Thanks,
Dmitri

**Comment by siying** [comment]
Too late. I already reformatted...

**Comment by yuslepukhin** [comment]
OK, I will include those into the PR.

Dmitri
